src/main/java/com/visenze/productsearch/ProductSearch.java

```diff
@@ -0,0 +1,213 @@
+         *
+         * @see ProductSearch
+         */
+        private String endpoint = DEFAULT_ENDPOINT;
+
+        /**
+         * Client configurations such as time-outs etc. These configurations
+         * will be used to construct the ViHttpClient class that is stored as
+         * member variable of ProductSearch
+         *
+         * @see ClientConfig
+         * @see ProductSearch
+         */
+        private ClientConfig config;
+
+        /**
+         * Constructor with only highly necessary parameters.
+         *
+         * @param app_key unique access key used as credentials
+         * @param placement_id which placement ID (from dashboard)
+         */
+        public Builder(String app_key, Integer placement_id) {
+            this.appKey      = app_key;
+            this.placementId = placement_id;
+            this.config       = new ClientConfig();
+        }
+
+        /**
+         * Build the ProductSearch class based on configurations set through the
+         * builder pattern.
+         *
+         * @return ProductSearch object class
+         */
+        public ProductSearch build() {
+            if (this.appKey == null || this.appKey.isEmpty()) {
+                throw new InternalViSearchException(ResponseMessages.INVALID_KEY);
+            }
+            return new ProductSearch(this.appKey, this.placementId, endpoint, config);
+        }
+
+        /**
+         * Set api custom url endpoint for
+         *
+         * @param url The custom url endpoint to use instead
+         *
+         * @return this 'itself'
+         */
+        public Builder setApiEndPoint(String url) {
+            this.endpoint = url;
+            return this;
+        }
+
+        /**
+         * Set custom client configurations for http connections
+         *
+         * @param config custom client http configurations
+         *
+         * @return this 'itself'
+         *
+         * @see ClientConfig
+         */
+        public Builder setClientConfig(ClientConfig config) {
+            this.config = config;
+            return this;
+        }
+    }
+
+    /**
+     * Constructor of ProductSearch made private to only allow it's Builder
+     * class to create it.
+     *
+     * @param appKey Unique app key that acts as authenticator for the client
+     * @param placementId Placement ID of the template that was chosen when
+     *                     creating apps on the dashboard
+     * @param endpoint Which endpoint to use for all ViHttpClient queries
+     * @param config Configuration to the behaviours of the ViHttpClient
+     */
+    private ProductSearch(String appKey, Integer placementId, String endpoint,
+                          ClientConfig config)
+    {
+        this.appKey       = appKey;
+        this.placementId  = placementId;
+        this.endpoint     = endpoint;
+        this.httpClient   = new ViSearchHttpClientImpl(this.endpoint, appKey, placementId.toString(), config);
```

Comment:
need to create a ProductSearchHttpClientImpl that extends from ViSearchHttpClientImpl. Placement ID is not a secret key. Cannot just reuse like this. It just doesnt make sense. For the private method within ViSearchHttpClientImpl, if you want to reuse make them protected.

Reply:
ViSearchHttpClientImpl by default will generate a Basic Authentication header for access + secret key. This is not applicable to Product Search API.